# Cubo Gelatinoso
Heloísa Fernandes Cano - T26

## Enunciado

### Objetivo

<b>Objetivo:</b> Escolha um conjunto de dados que siga as considerações abaixo e estude o desempenho de modelos induzidos pelo algoritmo de <i>k</i> vizinhos mais próximos neste conjunto de dados 
considerando diferentes conjuntos de hiperparâmetros (pelo menos 10 conjuntos de hiperparâmetros 
diferentes). A entrega deve ser realizada em um notebook de Jupyter. A sugestão é que esta seja uma 
entrega com texto direto, sem a necessidade de longos parágrafos. É necessário utilizar o `scikit‑learn`
para induzir os modelos de <i>k</i> vizinhos mais próximos. Não se esqueça de justificar a métrica de 
desempenho escolhida.

### Considerações

<b>Considerações:</b> O conjunto de dados deve ter as seguintes características:
- Pelo menos 5 atributos. Destes atributos, pelo menos um deles deve ser categórico
- Pelo menos 1 target (não há necessidade de mais que 1 target e pode ser um target categórico 
ou numérico, a sua escolha)
- Pelo menos 150 exemplos
- Os dados não podem ser sintéticos


### Expectativas

<b>Expectativas:</b> É esperado que o trabalho siga as boas práticas apresentadas na disciplina, apresente 
resultados de forma gráfica, use seções bem nomeadas e bem divididas no notebook de Jupyter e 
encerre a entrega com uma seção chamada “Conclusões e principais aprendizados”. O trabalho deve 
ter uma seção chamada “Descrição do uso de IA neste trabalho” detalhando como foi a interação com 
ferramentas de IA para a construção dos textos e códigos.

### Pontuação base

<b>Pontuação base:</b> 7 (válida apenas caso o objetivo tenha sido realizado)
Descontos usuais da pontuação base:
- Alguma expectativa não foi atingida
- Trabalho não seguiu boas práticas apresentadas na disciplina. Lembre‑se de revisar o material 
apresentado e suas anotações durante a aula
- Trabalho não apresentou resultados de forma gráfica
- Não guiou o leitor na entrega
- Não interpretou e discutiu os resultados
- Gráfico sem legenda em um ou mais eixos
- Código que acusa erro ao rodar na ordem de cima para baixo
- Entrega sem referências
- Uso incorreto ou omisso de citações (incluindo falta de citação sobre uso de IA)
- Não soube explicar sobre estratégias e códigos utilizados na entrega
- Respostas incorretas às perguntas realizadas pelo professor durante o momento avaliativo
- Respostas incorretas às perguntas realizadas pelo Perguntador 2000™ durante o momento 
avaliativo
- Atraso na entrega do trabalho no Teams

### Pontuação complementar

<b>Pontuação complementar:</b> 3 (cada item abaixo pode valer até no máximo 1 ponto)
- Realizou uma boa análise exploratória dos dados e comparou o desempenho dos modelos com 
relação a um modelo baseline
- Checou como o desempenho dos modelos varia utilizando ou não utilizando dados normalizados de diferentes maneiras
- Checou o desempenho dos modelos variando o número de vizinhos
- Checou o desempenho dos modelos variando a função de distância (considerando pelo menos 
3 funções diferentes)
- Checou o desempenho dos modelos variando a estratégia de codificação do(s) atributo(s) categórico(s)



### Observações e Lembretes

<b>Motivação e aprendizado:</b> Esta é uma tarefa introdutória da disciplina, sendo assim, a motivação 
principal é justamente iniciar sua jornada no campo do Aprendizado de Máquina.

<b>Sugestão de uso de IA nesta tarefa:</b> Um bom uso de IA pode ser para encontrar conjuntos 
de dados de seu interesse. Fora isso, debater com a IA sobre quais atividades complementares fazer, 
ponderando aprendizado e interesse, também é uma boa estratégia.

<b>Lembretes:</b>
- “Guiar o leitor” não é sinônimo de “escrever um livro”. É perfeitamente possível guiar o leitor 
com textos curtos e diretos. Um texto que guia o leitor é aquele que deixa claro o que será 
realizado antes de realizar a tarefa e depois discute o que foi observado.
- Citações devem aparecer o mais próximo possível do local onde a informação foi apresentada 
ao leitor. Ter apenas uma seção de referências sem as devidas citações no texto não é a maneira 
correta de redigir um texto científico e implicará em descontos na pontuação.

## Introdução

### Importando bibliotecas

In [1]:
import pandas as pd
import random as rd

semente_aleatoria = 1225

### Lendo o arquivo

A primeira coisa que precisamos fazer é realizar a leitura do arquivo que vamos utilizar. Utilizamos o método `pd.read_csv()` para ler o arquivo `combined.csv`. É importante que o arquivo csv esteja na mesma pasta que esse notebook!!

In [2]:
arquivo = "combined.csv"
df = pd.read_csv(arquivo, sep=",", header=None)
df

,0,1,2,3,4,5,6,7,8,9,10,11,12
0,record_id,month,day,year,plot_id,species_id,sex,hindfoot_length,weight,genus,species,taxa,plot_type
1,1,7,16,1977,2,NL,M,32,NaN,Neotoma,albigula,Rodent,Control
2,72,8,19,1977,2,NL,M,31,NaN,Neotoma,albigula,Rodent,Control
3,224,9,13,1977,2,NL,NaN,NaN,NaN,Neotoma,albigula,Rodent,Control
4,266,10,16,1977,2,NL,NaN,NaN,NaN,Neotoma,albigula,Rodent,Control
...,...,...,...,...,...,...,...,...,...,...,...,...,...
34782,26966,10,25,1997,7,PL,M,20,16,Peromyscus,leucopus,Rodent,Rodent Exclosure
34783,27185,11,22,1997,7,PL,F,21,22,Peromyscus,leucopus,Rodent,Rodent Exclosure
34784,27792,5,2,1998,7,PL,F,20,8,Peromyscus,leucopus,Rodent,Rodent Exclosure
34785,28806,11,21,1998,7,PX,NaN,NaN,NaN,Chaetodipus,sp.,Rodent,Rodent Exclosure


### Tratamento e filtro de dados

Agora que lemos o arquivo, podemos perceber que as colunas estão com números como seus nomes, e não categorias. Por outro lado, as categorias estão presentes na primeira linha do dataframe. Então, vamos renomear as colunas de acordo com a primeira linha, utilizando o método `.rename()`.\

In [3]:
df = df.rename(columns={0: 'record_id', 1: 'month', 2: 'day', 3: 'year', 4: 'plot_id', 5: 'species_id', 6: 'sex', 
                   7: 'hindfoot_lenght', 8: 'weight', 9: 'genus', 10: 'species', 11: 'taxa', 12: 'plot_type'
                  })
df

,record_id,month,day,year,plot_id,species_id,sex,hindfoot_lenght,weight,genus,species,taxa,plot_type
0,record_id,month,day,year,plot_id,species_id,sex,hindfoot_length,weight,genus,species,taxa,plot_type
1,1,7,16,1977,2,NL,M,32,NaN,Neotoma,albigula,Rodent,Control
2,72,8,19,1977,2,NL,M,31,NaN,Neotoma,albigula,Rodent,Control
3,224,9,13,1977,2,NL,NaN,NaN,NaN,Neotoma,albigula,Rodent,Control
4,266,10,16,1977,2,NL,NaN,NaN,NaN,Neotoma,albigula,Rodent,Control
...,...,...,...,...,...,...,...,...,...,...,...,...,...
34782,26966,10,25,1997,7,PL,M,20,16,Peromyscus,leucopus,Rodent,Rodent Exclosure
34783,27185,11,22,1997,7,PL,F,21,22,Peromyscus,leucopus,Rodent,Rodent Exclosure
34784,27792,5,2,1998,7,PL,F,20,8,Peromyscus,leucopus,Rodent,Rodent Exclosure
34785,28806,11,21,1998,7,PX,NaN,NaN,NaN,Chaetodipus,sp.,Rodent,Rodent Exclosure


Agora, vamos utilizar o método `.dropna()` para remover as linhas que tiverem informações faltando.

In [4]:
df = df.dropna()
df

,record_id,month,day,year,plot_id,species_id,sex,hindfoot_lenght,weight,genus,species,taxa,plot_type
0,record_id,month,day,year,plot_id,species_id,sex,hindfoot_length,weight,genus,species,taxa,plot_type
12,845,5,6,1978,2,NL,M,32,204,Neotoma,albigula,Rodent,Control
14,1164,8,5,1978,2,NL,M,34,199,Neotoma,albigula,Rodent,Control
15,1261,9,4,1978,2,NL,M,32,197,Neotoma,albigula,Rodent,Control
18,1756,4,29,1979,2,NL,M,33,166,Neotoma,albigula,Rodent,Control
...,...,...,...,...,...,...,...,...,...,...,...,...,...
34780,26557,7,29,1997,7,PL,F,20,22,Peromyscus,leucopus,Rodent,Rodent Exclosure
34781,26787,9,27,1997,7,PL,F,21,16,Peromyscus,leucopus,Rodent,Rodent Exclosure
34782,26966,10,25,1997,7,PL,M,20,16,Peromyscus,leucopus,Rodent,Rodent Exclosure
34783,27185,11,22,1997,7,PL,F,21,22,Peromyscus,leucopus,Rodent,Rodent Exclosure


Como nós vamos trabalhar apenas com roedores, vamos filtrar o data frame para que sejam exibidas apenas as linhas cuja coluna `"taxa"` seja igual a `"Rodent"`, com o método `.loc[]`.

In [5]:
logica = df['taxa'] == "Rodent"
df = df.loc[logica]
df

,record_id,month,day,year,plot_id,species_id,sex,hindfoot_lenght,weight,genus,species,taxa,plot_type
12,845,5,6,1978,2,NL,M,32,204,Neotoma,albigula,Rodent,Control
14,1164,8,5,1978,2,NL,M,34,199,Neotoma,albigula,Rodent,Control
15,1261,9,4,1978,2,NL,M,32,197,Neotoma,albigula,Rodent,Control
18,1756,4,29,1979,2,NL,M,33,166,Neotoma,albigula,Rodent,Control
19,1818,5,30,1979,2,NL,M,32,184,Neotoma,albigula,Rodent,Control
...,...,...,...,...,...,...,...,...,...,...,...,...,...
34780,26557,7,29,1997,7,PL,F,20,22,Peromyscus,leucopus,Rodent,Rodent Exclosure
34781,26787,9,27,1997,7,PL,F,21,16,Peromyscus,leucopus,Rodent,Rodent Exclosure
34782,26966,10,25,1997,7,PL,M,20,16,Peromyscus,leucopus,Rodent,Rodent Exclosure
34783,27185,11,22,1997,7,PL,F,21,22,Peromyscus,leucopus,Rodent,Rodent Exclosure


Antes de decidirmos quais serão as duas espécies que vamos prever, precisamos saber quantos dados de cada uma dessas espécies existem na tabela. Para isso, vamos utilizar o método `.values_counts()` na coluna `species`, para que seja possível visualizar quantos exemplares de cada espécie foram registrados.

In [6]:
serie_species = pd.Series( df["species"] )
frequencia_species = serie_species.value_counts()#.sort_index()
print(frequencia_species)

species
merriami        9727
penicillatus    2969
baileyi         2803
ordii           2790
megalotis       2417
torridus        2081
spectabilis     2023
flavus          1469
eremicus        1198
albigula        1045
leucogaster      905
maniculatus      835
hispidus         159
fulvescens        73
taylori           45
ochrognathus      40
fulviventer       38
leucopus          35
sp.                9
montanus           8
intermedius        7
Name: count, dtype: int64


Queremos que as espécies que vamos tentar prever tenham pelo menos 150 exemplares cada uma. Vamos filtrar o data frame novamente para que atenda esse requisito:

In [7]:
# localizamos quais sao as especies que têm pelo menos de 150 exemplos
especies_desejadas = []
for especie, quantidade in frequencia_species.items():
    if quantidade >= 150:
        especies_desejadas.append( especie )

print(especies_desejadas)

['merriami', 'penicillatus', 'baileyi', 'ordii', 'megalotis', 'torridus', 'spectabilis', 'flavus', 'eremicus', 'albigula', 'leucogaster', 'maniculatus', 'hispidus']


Utilizamos o método `isin()` [2] para filtrar apenas as espécies que estejam dentro da lista de "espécies desejadas".

In [8]:
logica = df['species'].isin( especies_desejadas )
df = df.loc[logica]
df

,record_id,month,day,year,plot_id,species_id,sex,hindfoot_lenght,weight,genus,species,taxa,plot_type
12,845,5,6,1978,2,NL,M,32,204,Neotoma,albigula,Rodent,Control
14,1164,8,5,1978,2,NL,M,34,199,Neotoma,albigula,Rodent,Control
15,1261,9,4,1978,2,NL,M,32,197,Neotoma,albigula,Rodent,Control
18,1756,4,29,1979,2,NL,M,33,166,Neotoma,albigula,Rodent,Control
19,1818,5,30,1979,2,NL,M,32,184,Neotoma,albigula,Rodent,Control
...,...,...,...,...,...,...,...,...,...,...,...,...,...
34775,32477,8,25,2001,7,PB,M,28,32,Chaetodipus,baileyi,Rodent,Rodent Exclosure
34776,33103,11,17,2001,7,PB,M,28,41,Chaetodipus,baileyi,Rodent,Rodent Exclosure
34777,33305,12,15,2001,7,PB,M,29,44,Chaetodipus,baileyi,Rodent,Rodent Exclosure
34778,34524,7,13,2002,7,PB,M,25,16,Chaetodipus,baileyi,Rodent,Rodent Exclosure


Para vizualizarmos novamente quais espécies ainda estão no data frame e quantos exemplares existem de cada uma, utilizamos o método `.value_counts()` novamente.

In [9]:
serie_species = pd.Series( df["species"] )
frequencia_species = serie_species.value_counts()
print(frequencia_species)

species
merriami        9727
penicillatus    2969
baileyi         2803
ordii           2790
megalotis       2417
torridus        2081
spectabilis     2023
flavus          1469
eremicus        1198
albigula        1045
leucogaster      905
maniculatus      835
hispidus         159
Name: count, dtype: int64


Como ainda são muitas espécies e precisamos de apenas duas para fazer uma previsão de classificação binária, vamos escolher duas espécies. As espécies escolhidas foram a espécie `ordii`, com 2790 exemplares, e a espécie `spectabilis`, com 2023 exemplares. Vamos filtrar o data frame mais uma vez.

In [10]:
df = df.loc[ df['species'].isin( ['ordii', 'spectabilis'] ) ]
df

,record_id,month,day,year,plot_id,species_id,sex,hindfoot_lenght,weight,genus,species,taxa,plot_type
949,585,2,18,1978,2,DS,M,53,157,Dipodomys,spectabilis,Rodent,Control
950,657,3,11,1978,2,DS,M,54,153,Dipodomys,spectabilis,Rodent,Control
951,731,4,8,1978,2,DS,F,47,110,Dipodomys,spectabilis,Rodent,Control
953,835,5,6,1978,2,DS,M,53,149,Dipodomys,spectabilis,Rodent,Control
955,934,5,19,1978,2,DS,F,47,114,Dipodomys,spectabilis,Rodent,Control
...,...,...,...,...,...,...,...,...,...,...,...,...,...
34364,33567,2,10,2002,7,DO,M,36,62,Dipodomys,ordii,Rodent,Rodent Exclosure
34365,33568,2,10,2002,7,DO,F,37,55,Dipodomys,ordii,Rodent,Rodent Exclosure
34366,33569,2,10,2002,7,DO,F,38,47,Dipodomys,ordii,Rodent,Rodent Exclosure
34367,33570,2,10,2002,7,DO,F,35,54,Dipodomys,ordii,Rodent,Rodent Exclosure


### Sorteando dados

Porém, ainda temos uma certa diferença na quantidade de dados: existem 2790 exemplares da espécie ordii e 2023 exemplares da espécie spectabilis, com uma diferença de 767. Para que isso não influencie no modelo, vamos sortear 2000 dados de cada uma dessas espécies. [3] semente aleatoria [4] juntar dataframes [5]

In [11]:
df_ordii = df.loc[ df["species"] == "ordii" ]
df_spectabilis = df.loc[ df["species"] == 'spectabilis' ]

ordii_aleatorios = df_ordii.sample(2000, random_state = semente_aleatoria)
spectabilis_aleatorios = df_spectabilis.sample(2000, random_state = semente_aleatoria)

df = pd.merge(ordii_aleatorios, spectabilis_aleatorios, how = 'outer')
df

,record_id,month,day,year,plot_id,species_id,sex,hindfoot_lenght,weight,genus,species,taxa,plot_type
0,1000,6,9,1978,5,DS,F,46,71,Dipodomys,spectabilis,Rodent,Rodent Exclosure
1,10002,2,17,1985,9,DO,M,35,51,Dipodomys,ordii,Rodent,Spectab exclosure
2,10012,2,17,1985,9,DO,F,34,36,Dipodomys,ordii,Rodent,Spectab exclosure
3,10015,2,17,1985,5,DO,F,32,35,Dipodomys,ordii,Rodent,Rodent Exclosure
4,10024,2,17,1985,1,DO,F,34,59,Dipodomys,ordii,Rodent,Spectab exclosure
...,...,...,...,...,...,...,...,...,...,...,...,...,...
3995,9977,2,17,1985,4,DO,M,34,43,Dipodomys,ordii,Rodent,Control
3996,998,6,9,1978,14,DS,F,51,122,Dipodomys,spectabilis,Rodent,Control
3997,9981,2,17,1985,9,DS,F,48,120,Dipodomys,spectabilis,Rodent,Spectab exclosure
3998,9983,2,17,1985,9,DS,M,48,81,Dipodomys,spectabilis,Rodent,Spectab exclosure


## Uso de inteligência artificial

## Referências

- [2] https://labex.io/pt/tutorials/pandas-pandas-filtering-data-596393
- [3] https://pt.stackoverflow.com/questions/324588/fazer-um-random-de-um-dataset-com-pandas
- [4] https://stackoverflow.com/questions/55748741/how-to-make-a-seed-to-pd-sample-like-np-random-seed
- [5] https://pt.stackoverflow.com/questions/515603/juntar-dois-dataframes-de-mesmo-nome-mantendo-todas-as-colunas